google/cloud/spanner/value_test.cc

```diff
@@ -201,6 +214,114 @@ TEST(Value, SpannerArray) {
   EXPECT_FALSE(null_vd.get<ArrayInt64>().ok());
 }
 
+TEST(Value, SpannerStruct) {
+  // Using declarations to shorten the tests making them more readable.
+  using std::int64_t;
+  using std::make_pair;
+  using std::make_tuple;
+  using std::pair;
+  using std::string;
+  using std::tuple;
+
+  auto tup1 = make_tuple(false, int64_t{123});
+  using T1 = decltype(tup1);
+  Value v1(tup1);
+  EXPECT_TRUE(v1.is<T1>());
+  EXPECT_FALSE(v1.is_null<T1>());
+  EXPECT_TRUE(v1.get<T1>().ok());
+  EXPECT_EQ(tup1, *v1.get<T1>());
+  EXPECT_EQ(v1, v1);
+
+  // Verify that wrapping an element in a pair doesn't affect its C++ type.
+  EXPECT_TRUE((v1.is<tuple<bool, int64_t>>()));
+  EXPECT_TRUE((v1.is<tuple<pair<string, bool>, int64_t>>()));
+  EXPECT_TRUE((v1.is<tuple<bool, pair<string, int64_t>>>()));
+  EXPECT_TRUE((v1.is<tuple<pair<string, bool>, pair<string, int64_t>>>()));
+
+  auto tup2 = make_tuple(false, make_pair(string("f2"), int64_t{123}));
+  using T2 = decltype(tup2);
+  Value v2(tup2);
+  EXPECT_TRUE(v2.is<T2>());
+  EXPECT_FALSE(v2.is_null<T2>());
+  EXPECT_TRUE(v2.get<T2>().ok());
+  EXPECT_EQ(tup2, *v2.get<T2>());
+  EXPECT_EQ(v2, v2);
+  EXPECT_NE(v2, v1);
+
+  // T1 is lacking field names, but otherwise the same as T2.
+  EXPECT_EQ(tup1, *v2.get<T1>());
+  EXPECT_NE(tup2, *v1.get<T2>());
+
+  auto tup3 = make_tuple(false, make_pair(string("Other"), int64_t{123}));
+  using T3 = decltype(tup3);
+  Value v3(tup3);
+  EXPECT_TRUE(v3.is<T3>());
+  EXPECT_FALSE(v3.is_null<T3>());
+  EXPECT_TRUE(v3.get<T3>().ok());
+  EXPECT_EQ(tup3, *v3.get<T3>());
+  EXPECT_EQ(v3, v3);
+  EXPECT_NE(v3, v2);
+  EXPECT_NE(v3, v1);
+
+  static_assert(std::is_same<T2, T3>::value, "Only diff is field name");
+
+  // v1 != v2, yet T2 works with v1 and vice versa
+  EXPECT_NE(v1, v2);
+  EXPECT_TRUE(v1.is<T2>());
+  EXPECT_FALSE(v1.is_null<T2>());
+  EXPECT_TRUE(v2.is<T1>());
+  EXPECT_FALSE(v2.is_null<T1>());
+
+  Value v_null(optional<T1>{});
+  EXPECT_TRUE(v_null.is<T1>());
+  EXPECT_TRUE(v_null.is<T2>());
+  EXPECT_TRUE(v_null.is_null<T1>());
+  EXPECT_TRUE(v_null.is_null<T2>());
+
+  EXPECT_NE(v1, v_null);
+  EXPECT_NE(v2, v_null);
+
+  auto array_struct = std::vector<T3>{
+      {false, {"age", 1}},
+      {true, {"age", 2}},
+      {false, {"age", 3}},
+  };
```

Comment:
It looks like gcc 4.8 is having trouble with these T3 initializers.
```
error: converting to 'std::tuple<bool, std::pair<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, long int> >' from initializer list would use explicit constructor 'constexpr std::tuple<_T1, _T2>::tuple(const _T1&, const _T2&) [with _T1 = bool; _T2 = std::pair<std::basic_string<char>, long int>]'
```
Perhaps wrapping them in T3{} is sufficient?